operational_analysis/methods/plant_analysis.py

```diff
@@ -776,12 +777,24 @@ def run_regression(self, n):
         # Machine learning models
         else: 
             ml = MachineLearningSetup(self.reg_model, **self.ml_setup_kwargs)
-            ml.hyper_optimize(np.array(reg_data[:,0:-1]), reg_data[:,-1], n_iter_search = 5, report = False, cv = KFold(n_splits = 2))
-            predicted_y = ml.random_search.predict(np.array(reg_data[:,0:-1]))
+            # Memoized approach for optimized hyperparameters
+            if self._run.reanalysis_product in self.opt_model:
+                self.opt_model[(self._run.reanalysis_product)].fit(np.array(reg_data[:,0:-1]), reg_data[:,-1])
+                predicted_y = self.opt_model[(self._run.reanalysis_product)].predict(np.array(reg_data[:,0:-1]))
```

Comment:
Here `fit` and `predict` are being called from self.opt_model, which is defined in `machine_learning_setup` as `self.opt_model = self.random_search.best_estimator_`. But below after the `else` statement, `predict` is called from `ml.random_search`. Also, within machine_learning_setup the model is fit using `self.random_search.fit`. Why are `fit` and `predict` being called using `ml.random_search.best_estimator_.fit` in one case and `ml.random_search.fit` in the other? Is there a reason for saving the opt_model as `self.random_search.best_estimator_` instead of just `self.random_search` in machine_learning_setup?

Reply:
Thanks for catching this inconsistency. 
I've re-factored this chunk of the code, so that now predict is always called using the `best_estimator_`.
In machine_learning_setup, it is correct to call fit on `ml.random_search`, as `best_estimator_` becomes available only after `ml.random_search` is fitted first.
Basically, the idea is: fit random_search only once for each reanalysis product, and then use its `best_estimator_` for all the rest of the calculations.

Reply:
Got it, that makes more sense and the code looks a lot cleaner, especially removing some duplication. One last question, when the `opt_model` is already found in the memoized dictionary, you are still calling `fit` on the saved `best_estimator_`. Is that step necessary? Does `ml.random_search.fit` set the hyperparameters and model parameters, and `ml.random_search.best_estimator_` just set the model parameters for the specific data sample? Just trying to understand where the time savings comes from if `fit` is still called.

Reply:
Yes, the additional fit is needed to set the model weights to the specific data used at each MC iteration, but not the hyperparameters, which are taken from the `opt_model`. So the time savings come from skipping the cross validation needed to set the hyperparameters, so that only the fitting of the model weights is performed at each iteration.